services/twofa.go

```diff
@@ -0,0 +1,283 @@
+	RecoveryKey *string `json:"recoveryKey,omitempty" validate:"required_without=TOTPCode,excluded_with=TOTPCode"`
+}
+
+// TwoFAService provides methods for managing two-factor authentication
+type TwoFAService struct {
+	// issuer is the name of the issuer for TOTP keys
+	issuer string
+	// qrSize is the size in pixels for generated QR codes
+	qrSize int
+	// ds is the datastore instance for persistent storage
+	ds *datastore.Datastore
+	// keyServiceClient is the client used to communicate with the key service
+	keyServiceClient *util.KeyServiceClient
+	// isKeyService indicates whether this instance is the key service
+	isKeyService bool
+}
+
+// NewTwoFAService creates a new TwoFAService instance with configuration from environment
+func NewTwoFAService(ds *datastore.Datastore, isKeyService bool) *TwoFAService {
+	issuer := os.Getenv(totpIssuerEnv)
+	if issuer == "" {
+		issuer = defaultIssuer
+	}
+
+	qrSize := defaultQRSize
+	if sizeStr := os.Getenv(totpQRSizeEnv); sizeStr != "" {
+		size, err := strconv.Atoi(sizeStr)
+		if err != nil || size <= 0 {
+			log.Panic().Err(err).Msgf("invalid TOTP QR size: %s", sizeStr)
+		}
+		qrSize = size
+	}
+
+	// Create a client if we're not the key service and KEY_SERVICE_URL is set
+	var client *util.KeyServiceClient
+	if !isKeyService && os.Getenv(util.KeyServiceURLEnv) != "" {
+		client = util.NewKeyServiceClient()
+	}
+
+	return &TwoFAService{
+		issuer:           issuer,
+		qrSize:           qrSize,
+		ds:               ds,
+		keyServiceClient: client,
+		isKeyService:     isKeyService,
+	}
+}
+
+// DisableTwoFA disables two-factor authentication for an account
+func (t *TwoFAService) DisableTwoFA(accountID uuid.UUID) error {
+	if err := t.ds.SetTOTPSetting(accountID, false); err != nil {
+		return err
+	}
+
+	if err := t.DeleteTOTPKey(accountID); err != nil {
+		return err
+	}
+
+	if err := t.ds.SetRecoveryKey(accountID, nil); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// ProcessAuthRequest verifies either TOTP code or recovery key for an account
+func (t *TwoFAService) ProcessAuthRequest(loginState *datastore.InterimPasswordState, req *TwoFAAuthRequest) error {
+	// Verify either TOTP code or recovery key
+	if req.TOTPCode != nil {
+		// Verify TOTP code
+		if err := t.ValidateTOTPCode(*loginState.AccountID, *req.TOTPCode); err != nil {
+			return err
+		}
+	} else if req.RecoveryKey != nil {
+		// Verify recovery key
+		if err := t.ds.CheckRecoveryKey(*loginState.AccountID, *req.RecoveryKey); err != nil {
+			return err
+		}
+		if err := t.DisableTwoFA(*loginState.AccountID); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// GenerateAndStoreTOTPKey creates and stores a new TOTP key for an account
+func (t *TwoFAService) GenerateAndStoreTOTPKey(accountID uuid.UUID, email string) (*otp.Key, error) {
+	if t.keyServiceClient != nil {
+		return t.makeKeyServiceTOTPGenerateRequest(accountID, email)
+	}
+
+	// Otherwise, generate and store the key locally
+	opts := totp.GenerateOpts{
```

Comment:
Sadly, we can't touch the `SecretSize` (or Google Authenticator for iOS will refuse to use the QR code) or the `Algorithm` (or lots of broken authenticators like Authy, Duo and Microsoft will happily accept the code and generate bad 2FA codes).

i.e. No changes needed here